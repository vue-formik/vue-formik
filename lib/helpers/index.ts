import updateNestedProperty from "@/helpers/updateNestedProperty";
import getNestedValue from "@/helpers/getNestedValue";
import clearReactiveObject from "@/helpers/clearReactiveObject";
import { constructLabel } from "@/helpers/utils";

export {
  updateNestedProperty,
  getNestedValue,
  clearReactiveObject,
  constructLabel,
};
