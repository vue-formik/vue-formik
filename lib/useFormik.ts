import { computed, reactive, toRaw, watch } from 'vue'
import { ObjectSchema } from 'yup'
import type { FormikHelpers, IUseFormik, ValidationRule } from '@/types'
import { getNestedValue, updateNestedProperty } from '@/helpers'

const useFormik = <T extends object>({
  initialValues,
  validationSchema,
  onSubmit,
}: IUseFormik<T>) => {
  const validate = (values: T): Partial<Record<keyof T, string>> => {
    const validationErrors: Partial<Record<keyof T, string>> = {};

    if (validationSchema instanceof ObjectSchema) {
      try {
        const vSchema = validationSchema as ObjectSchema<T>;
        vSchema.validateSync(values, { abortEarly: false });
      } catch (e) {
        const err = e as never as { inner: { path: string; message: string }[] };
        if (typeof err === "object" && Array.isArray(err?.inner)) {
          err.inner.forEach((error) => {
            validationErrors[error.path as keyof T] = error.message;
          });
        }
      }
    } else {
      for (const key in validationSchema) {
        const value = values[key as keyof T];
        const rule = validationSchema[key as keyof T] as ValidationRule<T[keyof T]>;

        if (typeof rule === 'function') {
          const error = rule(value);
          if (error) {
            validationErrors[key as keyof T] = error;
          }
        }
      }
    }

    return validationErrors;
  };

  const values = reactive(initialValues);
  const errors = reactive(validate(initialValues));
  const touched = reactive({} as Partial<Record<keyof T, boolean | Record<string, boolean>>>);

  const setValues = (newValues: T) => {
    Object.assign(values, newValues); // Safely copy values into reactive object
    Object.assign(errors, validate(newValues)); // Validate updated values
  };

  const setErrors = (newErrors: T) => {
    Object.assign(errors, newErrors);
  };

  const reset = () => {
    Object.assign(values, initialValues);
    Object.assign(errors, validate(initialValues));
  };

  const setFieldValue = (field: string, value: unknown) => {
    updateNestedProperty(values as Record<string, unknown>, field, value);
  };

  const setFieldTouched = (field: string, touchedValue: boolean) => {
    updateNestedProperty(touched as Record<string, unknown>, field, touchedValue);
  }

  const handleSubmit = (e: Event) => {
    e.preventDefault();
    onSubmit(
      toRaw(values) as T,
      {
        setErrors,
        reset,
        setValues,
      } as FormikHelpers<T>
    );
  };

  const handleBlur = (e: FocusEvent) => {
    const fieldName = (e.target as HTMLInputElement).name as keyof T;
    setFieldTouched(fieldName.toString(), true);
  }
  const handleChange = (e: Event) => {
    const target = e.target as HTMLInputElement;
    const fieldName = target.name as keyof T;
    const value = target.type === 'checkbox' ? target.checked : target.value;

    setFieldValue(fieldName.toString(), value);
  }

  const isValid = computed(() => {
    return Object.keys(errors).length === 0;
  })

  const isDirty = computed(() => {
    return JSON.stringify(values) !== JSON.stringify(initialValues);
  })

  watch(
    () => values,
    () => {
      const validationErrors = validate(toRaw(values) as T);
      //@ts-expect-error Object.keys(errors) is not iterable
      Object.keys(errors).forEach((key) => delete errors[key]); // Clear existing errors
      Object.assign(errors, validationErrors); // Assign new errors
    },
    { deep: true }
  );


  const hasFieldError = (field: string) => {
    const errorValue = getNestedValue(errors, field);
    const touchedValue = getNestedValue(touched, field);
    return (errorValue !== undefined && !!touchedValue);
  }

  const getFieldError = (field: string) => {
    if (hasFieldError(field)) {
      return errors[field as keyof typeof errors];
    } else {
      return '';
    }
  }

  return {
    values,
    errors,
    touched,
    setValues,
    reset,
    setFieldValue,
    setFieldTouched,
    handleBlur,
    handleChange,
    handleSubmit,
    hasFieldError,
    getFieldError,
    isValid,
    isDirty
  };
};

export default useFormik;
