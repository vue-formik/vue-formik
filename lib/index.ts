import useFormik from "@/composables/useFormik";
import FormTextArea from "@/components/FormTextArea.vue";
import FormTextField from "@/components/FormTextField.vue";

export { FormTextArea, FormTextField, useFormik };
