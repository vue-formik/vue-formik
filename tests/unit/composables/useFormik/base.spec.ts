import { describe, expect, test } from "vitest";
import { useFormik } from "../../../../lib";
import { nextTick } from "vue";
import { initialValues, onSubmit, validationSchema } from "./fixtures";

describe("useFormik base features", async () => {
  test("should initialize form with initial values", () => {
    const { values } = useFormik({ initialValues, onSubmit });
    expect(values).toMatchSnapshot();
  });
  test("should set values", () => {
    const { setValues, values } = useFormik({ initialValues, onSubmit });
    setValues({ name: "Updated Name" });
    expect(values).toMatchSnapshot();
  });

  test("should set errors", () => {
    const { setErrors, errors } = useFormik({ initialValues, onSubmit });
    setErrors({ name: "Value must be less than 255 characters" });
    expect(errors).toMatchSnapshot();
  });

  test("should reset form", async () => {
    const { reset, values, touched, setFieldTouched, setFieldValue } = useFormik({
      initialValues,
      onSubmit,
    });

    setFieldTouched("name", true);
    setFieldTouched("email", true);
    setFieldValue("name", "Updated Name");
    setFieldValue("email", "updated@mail.cc");

    await nextTick();

    expect(values).toMatchSnapshot();
    expect(touched).toMatchSnapshot();

    reset();

    await nextTick();

    expect(values).toMatchSnapshot();
    expect(touched).toMatchSnapshot();
  });

  test("should set field value", () => {
    const { setFieldValue, values } = useFormik({ initialValues, onSubmit });
    setFieldValue("name", "Updated Name");
    expect(values).toMatchSnapshot();
  });

  test("should set field touched", () => {
    const { setFieldTouched, touched } = useFormik({ initialValues, onSubmit });
    setFieldTouched("name", true);
    expect(touched).toMatchSnapshot();
  });

  describe("status", () => {
    test("'isValid' should return true when form is valid", async () => {
      const { isValid, handleFieldChange, handleFieldBlur } = useFormik({
        initialValues,
        onSubmit,
        validationSchema,
      });
      expect(isValid.value).toBe(true);

      handleFieldChange({ target: { name: "email", value: "Kiran" } });
      handleFieldBlur({ target: { name: "email" } });

      await nextTick();

      expect(isValid.value).toBe(false);
    });
    test("'isDirty' should return true when form is dirty", async () => {
      const { isDirty, handleFieldChange, handleFieldBlur, reset } = useFormik({
        initialValues,
        onSubmit,
        validationSchema,
      });
      expect(isDirty.value).toBe(false);

      handleFieldChange({ target: { name: "email", value: "Kiran" } });
      handleFieldBlur({ target: { name: "email" } });

      await nextTick();

      expect(isDirty.value).toBe(true);

      reset();

      await nextTick();

      expect(isDirty.value).toBe(false);
    });
  });
});
