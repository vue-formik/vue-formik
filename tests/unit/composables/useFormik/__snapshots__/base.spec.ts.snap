// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useFormik base features > should initialize form with initial values 1`] = `
{
  "email": "kiran@parajuli.cc",
  "name": "Kiran",
}
`;

exports[`useFormik base features > should reset form 1`] = `
{
  "email": "updated@mail.cc",
  "name": "Updated Name",
}
`;

exports[`useFormik base features > should reset form 2`] = `
{
  "email": true,
  "name": true,
}
`;

exports[`useFormik base features > should reset form 3`] = `
{
  "email": "kiran@parajuli.cc",
  "name": "Kiran",
}
`;

exports[`useFormik base features > should reset form 4`] = `{}`;

exports[`useFormik base features > should set errors 1`] = `
{
  "name": "Value must be less than 255 characters",
}
`;

exports[`useFormik base features > should set field touched 1`] = `
{
  "name": true,
}
`;

exports[`useFormik base features > should set field value 1`] = `
{
  "email": "kiran@parajuli.cc",
  "name": "Updated Name",
}
`;

exports[`useFormik base features > should set values 1`] = `
{
  "email": "kiran@parajuli.cc",
  "name": "Updated Name",
}
`;

exports[`useFormik base features > validate on mount > should validate on mount when turned on 1`] = `
{
  "email": "Email is required",
  "name": "This field is required",
}
`;
